Description: When gnutls_global_init() is called manually from the application
 check the urandom fd for validity.
 .
 That addresses the issue where a server closes all open file descriptors
 and then calls gnutls_global_init().
 .
 This patch cherrypicks the following changes from 3.3.10:
 692a140b56027937ac517484b8a95fcae16a8f73
 3d112d4d2c32fecc80b7beff14717a2713052de5
 ca1d1efdf6767a00ccefd4e0c7d573d27e563592
 ecd9d7cfe7fb93945fb5605979e690da1791a63a
 684b825f5f78cc7ad1f61be232fd20ee0bc5b56f
 c999b03d66bbde418f533ade319178451ae27e79
 25d7fab329afd72979fc9aa2a258438ae498da01
 e8ffb7944037a45e6f2436a3906ce37e5ea97e3e
 c6b5716491f93881cde540b9fc3b8c00bc85278c
 1424a2fd977f6b6373c9ca7f76b715acedb21902
Author: Andreas Metzler <ametzler@debian.org>
Bug-Debian: https://bugs.debian.org/768841

diff --git a/doc/cha-gtls-app.texi b/doc/cha-gtls-app.texi
index 2dd5c85..b4932ca 100644
--- a/doc/cha-gtls-app.texi
+++ b/doc/cha-gtls-app.texi
@@ -271,8 +271,7 @@ library.
 @node Initialization
 @subsection Initialization
 
-GnuTLS must be initialized before it can be used.  The library is
-initialized on load; prior to 3.3.0 was initialized by calling @funcref{gnutls_global_init}. 
+The GnuTLS library is initialized on load; prior to 3.3.0 was initialized by calling @funcref{gnutls_global_init}. 
 The initialization typically enables CPU-specific acceleration, performs any required
 precalculations needed, opens any required system devices (e.g., /dev/urandom on Linux)
 and initializes subsystems that could be used later.
@@ -280,6 +279,11 @@ and initializes subsystems that could be used later.
 The resources allocated by the initialization process will be released 
 on library deinitialization, or explictly by calling @funcref{gnutls_global_deinit}.
 
+Note that during initialization file descriptors may be kept open by
+GnuTLS (e.g. /dev/urandom) on library load. Applications closing all unknown file
+descriptors must immediately call @funcref{gnutls_global_init}, after that, to
+ensure they don't disrupt GnuTLS' operation.
+
 @c In order to take advantage of the internationalization features in
 @c GnuTLS, such as translated error messages, the application must set
 @c the current locale using @code{setlocale} before initializing GnuTLS.
diff --git a/lib/crypto-backend.h b/lib/crypto-backend.h
index 7f4e1b6..09a7d92 100644
--- a/lib/crypto-backend.h
+++ b/lib/crypto-backend.h
@@ -77,6 +77,7 @@ typedef struct {
 
 typedef struct gnutls_crypto_rnd {
 	int (*init) (void **ctx);
+	int (*check) (void **ctx);
 	int (*rnd) (void *ctx, int level, void *data, size_t datasize);
 	void (*rnd_refresh) (void *ctx);
 	void (*deinit) (void *ctx);
diff --git a/lib/gnutls_global.c b/lib/gnutls_global.c
index b9ca0cb..2786626 100644
--- a/lib/gnutls_global.c
+++ b/lib/gnutls_global.c
@@ -207,6 +207,16 @@ int gnutls_global_init(void)
 
 	_gnutls_init++;
 	if (_gnutls_init > 1) {
+		if (_gnutls_init == 2 && _gnutls_init_ret == 0) {
+			/* some applications may close the urandom fd 
+			 * before calling gnutls_global_init(). in that
+			 * case reopen it */
+			ret = _gnutls_rnd_check();
+			if (ret < 0) {
+				gnutls_assert();
+				goto out;
+			}
+		}
 		ret = _gnutls_init_ret;
 		goto out;
 	}
diff --git a/lib/nettle/rnd-common.c b/lib/nettle/rnd-common.c
index 0a015d5..116a131 100644
--- a/lib/nettle/rnd-common.c
+++ b/lib/nettle/rnd-common.c
@@ -37,6 +37,10 @@
 #include <rnd-common.h>
 #include <hash-pjw-bare.h>
 
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
 /* gnulib wants to claim strerror even if it cannot provide it. WTF */
 #undef strerror
 
@@ -94,6 +98,11 @@ int _rnd_get_system_entropy_win32(void* rnd, size_t size)
 
 get_entropy_func _rnd_get_system_entropy = _rnd_get_system_entropy_win32;
 
+int _rnd_system_entropy_check(void)
+{
+	return 0;
+}
+
 int _rnd_system_entropy_init(void)
 {
 	int old;
@@ -127,7 +136,8 @@ void _rnd_system_entropy_deinit(void)
 #include <locks.h>
 #include "egd.h"
 
-static int device_fd = -1;
+static int _gnutls_urandom_fd = -1;
+static mode_t _gnutls_urandom_fd_mode = 0;
 
 static int _rnd_get_system_entropy_urandom(void* _rnd, size_t size)
 {
@@ -137,7 +147,7 @@ static int _rnd_get_system_entropy_urandom(void* _rnd, size_t size)
 	for (done = 0; done < size;) {
 		int res;
 		do {
-			res = read(device_fd, rnd + done, size - done);
+			res = read(_gnutls_urandom_fd, rnd + done, size - done);
 		} while (res < 0 && errno == EINTR);
 
 		if (res <= 0) {
@@ -168,7 +178,7 @@ int _rnd_get_system_entropy_egd(void* _rnd, size_t size)
 
 	for (done = 0; done < size;) {
 		res =
-		    _rndegd_read(&device_fd, rnd + done, size - done);
+		    _rndegd_read(&_gnutls_urandom_fd, rnd + done, size - done);
 		if (res <= 0) {
 			if (res < 0) {
 				_gnutls_debug_log("Failed to read egd.\n");
@@ -186,31 +196,53 @@ int _rnd_get_system_entropy_egd(void* _rnd, size_t size)
 
 get_entropy_func _rnd_get_system_entropy = NULL;
 
+int _rnd_system_entropy_check(void)
+{
+	int ret;
+	struct stat st;
+
+	ret = fstat(_gnutls_urandom_fd, &st);
+	if (ret < 0 || st.st_mode != _gnutls_urandom_fd_mode) {
+		return _rnd_system_entropy_init();
+	}
+	return 0;
+}
+
 int _rnd_system_entropy_init(void)
 {
-int old;
-	
-	device_fd = open("/dev/urandom", O_RDONLY);
-	if (device_fd < 0) {
+	int old;
+	struct stat st;
+
+	_gnutls_urandom_fd = open("/dev/urandom", O_RDONLY);
+	if (_gnutls_urandom_fd < 0) {
 		_gnutls_debug_log("Cannot open urandom!\n");
 		goto fallback;
 	}
 
-	old = fcntl(device_fd, F_GETFD);
+	old = fcntl(_gnutls_urandom_fd, F_GETFD);
 	if (old != -1)
-		fcntl(device_fd, F_SETFD, old | FD_CLOEXEC);
+		fcntl(_gnutls_urandom_fd, F_SETFD, old | FD_CLOEXEC);
+
+	if (fstat(_gnutls_urandom_fd, &st) >= 0) {
+		_gnutls_urandom_fd_mode = st.st_mode;
+	}
 
 	_rnd_get_system_entropy = _rnd_get_system_entropy_urandom;
 
 	return 0;
 fallback:
-	device_fd = _rndegd_connect_socket();
-	if (device_fd < 0) {
+	_gnutls_urandom_fd = _rndegd_connect_socket();
+	if (_gnutls_urandom_fd < 0) {
 		_gnutls_debug_log("Cannot open egd socket!\n");
 		return
 			gnutls_assert_val
 			(GNUTLS_E_RANDOM_DEVICE_ERROR);
 	}
+
+	if (fstat(_gnutls_urandom_fd, &st) >= 0) {
+		_gnutls_urandom_fd_mode = st.st_mode;
+	}
+
 	_rnd_get_system_entropy = _rnd_get_system_entropy_egd;
 	
 	return 0;
@@ -218,9 +250,9 @@ fallback:
 
 void _rnd_system_entropy_deinit(void)
 {
-	if (device_fd >= 0) {
-		close(device_fd);
-		device_fd = -1;
+	if (_gnutls_urandom_fd >= 0) {
+		close(_gnutls_urandom_fd);
+		_gnutls_urandom_fd = -1;
 	}
 }
 #endif
diff --git a/lib/nettle/rnd-common.h b/lib/nettle/rnd-common.h
index a43f7be..c795f81 100644
--- a/lib/nettle/rnd-common.h
+++ b/lib/nettle/rnd-common.h
@@ -50,6 +50,7 @@ __attribute__((packed))
 void _rnd_get_event(struct event_st *e);
 
 int _rnd_system_entropy_init(void);
+int _rnd_system_entropy_check(void);
 void _rnd_system_entropy_deinit(void);
 
 typedef int (*get_entropy_func)(void* rnd, size_t size);
diff --git a/lib/nettle/rnd.c b/lib/nettle/rnd.c
index e8b0f40..d8c3c40 100644
--- a/lib/nettle/rnd.c
+++ b/lib/nettle/rnd.c
@@ -255,6 +255,15 @@ static int wrap_nettle_rnd_init(void **ctx)
 	return 0;
 }
 
+/* This is called when gnutls_global_init() is called for second time.
+ * It must check whether any resources are still available.
+ * The particular problem it solves is to verify that the urandom fd is still
+ * open (for applications that for some reason closed all fds */
+static int wrap_nettle_rnd_check(void **ctx)
+{
+	return _rnd_system_entropy_check();
+}
+
 static int
 wrap_nettle_rnd_nonce(void *_ctx, void *data, size_t datasize)
 {
@@ -363,6 +372,7 @@ int crypto_rnd_prio = INT_MAX;
 
 gnutls_crypto_rnd_st _gnutls_rnd_ops = {
 	.init = wrap_nettle_rnd_init,
+	.check = wrap_nettle_rnd_check,
 	.deinit = wrap_nettle_rnd_deinit,
 	.rnd = wrap_nettle_rnd,
 	.rnd_refresh = wrap_nettle_rnd_refresh,
diff --git a/lib/random.h b/lib/random.h
index 0aa1545..ee04549 100644
--- a/lib/random.h
+++ b/lib/random.h
@@ -48,4 +48,13 @@ inline static void _gnutls_rnd_refresh(void)
 void _gnutls_rnd_deinit(void);
 int _gnutls_rnd_init(void);
 
+inline static int _gnutls_rnd_check(void)
+{
+	return _gnutls_rnd_ops.check(gnutls_rnd_ctx);
+}
+
+#ifndef _WIN32
+extern int _gnutls_urandom_fd;
+#endif
+
 #endif
diff --git a/tests/Makefile.am b/tests/Makefile.am
index 522c01a..95d6541 100644
--- a/tests/Makefile.am
+++ b/tests/Makefile.am
@@ -84,7 +84,7 @@ ctests = mini-record-2 simple gc set_pkcs12_cred certder certuniqueid	\
 	 mini-cert-status mini-rsa-psk global-init sec-params \
 	 fips-test mini-global-load name-constraints x509-extensions \
 	 long-session-id mini-x509-callbacks-intr \
-	 crlverify
+	 crlverify init_fds
 
 if ENABLE_OCSP
 ctests += ocsp
diff --git a/tests/init_fds.c b/tests/init_fds.c
new file mode 100644
index 0000000..bf7a5de
--- /dev/null
+++ b/tests/init_fds.c
@@ -0,0 +1,80 @@
+/*
+ * Copyright (C) 2014 Nikos Mavrogiannopoulos
+ *
+ * Author: Nikos Mavrogiannopoulos
+ *
+ * This file is part of GnuTLS.
+ *
+ * GnuTLS is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * GnuTLS is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with GnuTLS; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA
+ */
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include <stdio.h>
+#include <unistd.h>
+#include <gnutls/gnutls.h>
+#include <gnutls/crypto.h>
+
+#include "utils.h"
+
+/* See <https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=760476>. */
+
+void doit(void)
+{
+#ifndef _WIN32
+	int res;
+	unsigned i;
+	int serial = 0;
+	char buf[128];
+
+	res = read(3, buf, 16);
+	if (res == 16)
+		serial = 1;
+
+	/* close all descriptors */
+	for (i=3;i<1024;i++)
+		close(i);
+
+	res = gnutls_global_init();
+	if (res != 0)
+		fail("global_init\n");
+
+	if (serial != 0) {
+		res = read(3, buf, 16);
+		if (res != 16) {
+			fail("could not open fd, or OS doesn't assign fds in a serial way (%d)\n", res);
+		}
+	}
+
+	res = gnutls_global_init();
+	if (res != 0)
+		fail("global_init2\n");
+
+	gnutls_rnd_refresh();
+
+	res = gnutls_rnd(GNUTLS_RND_RANDOM, buf, sizeof(buf));
+	if (res != 0)
+		fail("gnutls_rnd\n");
+
+	gnutls_global_deinit();
+
+	if (debug)
+		success("init-close success\n");
+#else
+	return;
+#endif
+}
